### ISSUE TITLE
Make it easier to delete properties/columns in 0-downtime scenarios 

### ISSUE THREAD

**Issue by MikaelEliasson** [dotnet/efcore#27165]
## The problem

Currently it's complicated to delete properties/columns in EF-core if you are trying to do 0-downtime deploys. Today it's easier to leave deprecated columns lying around than cleaning up. Which is bad for maintainability.

The reason is that EF generate queries with all columns named like this

```
SELECT [p].[Id], [p].[Name], [p].[PropToBeDeleted]
FROM [Persons] AS [p]
```

So when we migrate the DB during a release the old code will try to query a column that doesn't exist (unless we do projections everywhere). 

This is the best solution I've found so far https://github.com/MikaelEliasson/EFCoreSafelyDeleteColumnExample but that is still cumbersome and it also leaves the DB and the Model out of sync for a while because we manipulate the migration. From experience it's never a good idea to have them out of sync.

## Proposed solution

The best idea I have so far is something like

```
x.Entity<Person>().Property(p => p.PropToBeDeleted).IgnoreInQueries();
```

What this would do is to keep the property in the model so there is no migration to remove this column from the DB but the query above would be 

```
SELECT [p].[Id], [p].[Name]
FROM [Persons] AS [p]
```

**Comment by roji** [comment]
@MikaelEliasson the main problem I see with the above is that there's nothing stopping the (leftover) .NET property from being accidentally used; if that happens, the query would fail. I think removing the .NET property is a pretty important part of the process, to make sure no queries are using it any more.

> From experience it's never a good idea to have them out of sync.

Can you provide concrete information on this? What exactly is the problem with that?

Note that EF Core is generally fine with its model not being completely in sync with the database (in certain situations) - you could have additional tables in the database not managed by EF at all, or multiple tables of columns which were added externally, etc.

**Comment by MikaelEliasson** [comment]
@roji Ah, maybe it would be better to do the reverse then and remove the prop but say that it should still be there in the model.

When it comes why it's a bad idea here are a few reasons:
1) In tests you generally run with `db.EnsureCreated()` rather than apply migrations. You will test on different database schemas than in prod then, even if this particular case is unlikely to cause issues here.
2) Because of how the migrations folder grow, on any decent size project you will have to squash migrations. I think this approach https://www.bokio.se/engineering-blog/how-to-squash-ef-core-migrations/ will work even if you accidently pick a migration where the model is out of sync. But many people do this by cleaning out all migration and creating a new initial one. That would not work well if you time it badly.
3) When we migrated from EF6 to EF core some things where the Model and DB was out of sync caused quite a bit of grief. Solvable but still quite annoying. 
4) External tools, like bulk insert tools and similar will rely on the model. If it's out of sync there is a larger risk of issues.

In summary I think all issues are manageable. It would just be much significantly more pleasant if we could avoid them and I think it simplifies extensions to EF and similar.

**Comment by roji** [comment]
> Ah, maybe it would be better to do the reverse then and remove the prop but say that it should still be there in the model.

This is something you can easily do today, with [shadow properties](https://docs.microsoft.com/en-us/ef/core/modeling/shadow-properties), which are properties configured in the model which don't have a corresponding CLR property/field. Note that shadow *can* be accessed from queries (with EF.Property), so this doesn't completely prevent your application from accidentally depending on it, but it's pretty close.

> 1. In tests you generally run with db.EnsureCreated() rather than apply migrations. You will test on different database schemas than in prod then, even if this particular case is unlikely to cause issues here.

OK, but I don't really see the problem here... The new version of your application presumably don't use that property - so when running tests is neither in the model, nor in the database (as created from the model with EnsureCreated).

> 2. [...] many people do this by cleaning out all migration and creating a new initial one. That would not work well if you time it badly.

How so? Let's assume your database and model are out of sync, i.e. your database has an additional column that doesn't exist in the model. If you now clear out all migrations and recreate a single one, you're basically back to the same place - you have a single migration creating everything *without* that extra column (since the latest model version doesn't contain it), but your database still has it. So I don't see how squashing makes this any worse - what am I missing?

> 3. When we migrated from EF6 to EF core some things where the Model and DB was out of sync caused quite a bit of grief.

I can certainly believe that migration from EF6 to EF Core would cause lots of grief, for many reasons other than this sync issue. However, this is a one-time thing, and I don't think it should affect the larger question here: once you're using EF Core, you're just using EF Core.

> 4. External tools, like bulk insert tools and similar will rely on the model. If it's out of sync there is a larger risk of issues.

It would be good to understand exactly what that concretely means... if I'm understanding you right, all we have is some additional column in the DB which isn't in the model. If the tool is relying on the model, it should simply ignore that column - just as EF does. Of course, unknown external tools can have arbitrary behavior and I can't predict how they would react, but it would be good to at least have a concrete example or two.

In summary, I'm still not sure why one would one to keep the property in the model after stopping to use it. However, if this is what you want, shadow property seem like the right answer; I'm not sure if EF needs to provide anything else here.

**Comment by ajcvickers** [comment]
@MikaelEliasson Going back to your original scenario, can I clarify that the property is no longer used by newer versions of the application, but that you want the database column to be retained so that older versions of the application do not break?

**Comment by MikaelEliasson** [comment]
> @MikaelEliasson Going back to your original scenario, can I clarify that the property is no longer used by newer versions of the application, but that you want the database column to be retained so that older versions of the application do not break?

Exactly to be able to handle 0-downtime releases we need to be able to run the old code on the new version of the DB. This is also important in case of forced rollback of a deploy as it much nicer to work on the new version then do downgrade in that scenario.

**Comment by MikaelEliasson** [comment]
> > Ah, maybe it would be better to do the reverse then and remove the prop but say that it should still be there in the model.
> 
> This is something you can easily do today, with [shadow properties](https://docs.microsoft.com/en-us/ef/core/modeling/shadow-properties), which are properties configured in the model which don't have a corresponding CLR property/field. Note that shadow _can_ be accessed from queries (with EF.Property), so this doesn't completely prevent your application from accidentally depending on it, but it's pretty close.
> 
> > 1. In tests you generally run with db.EnsureCreated() rather than apply migrations. You will test on different database schemas than in prod then, even if this particular case is unlikely to cause issues here.
> 
> OK, but I don't really see the problem here... The new version of your application presumably don't use that property - so when running tests is neither in the model, nor in the database (as created from the model with EnsureCreated).
> 
> > 2. [...] many people do this by cleaning out all migration and creating a new initial one. That would not work well if you time it badly.
> 
> How so? Let's assume your database and model are out of sync, i.e. your database has an additional column that doesn't exist in the model. If you now clear out all migrations and recreate a single one, you're basically back to the same place - you have a single migration creating everything _without_ that extra column (since the latest model version doesn't contain it), but your database still has it. So I don't see how squashing makes this any worse - what am I missing?
> 
> > 3. When we migrated from EF6 to EF core some things where the Model and DB was out of sync caused quite a bit of grief.
> 
> I can certainly believe that migration from EF6 to EF Core would cause lots of grief, for many reasons other than this sync issue. However, this is a one-time thing, and I don't think it should affect the larger question here: once you're using EF Core, you're just using EF Core.
> 
> > 4. External tools, like bulk insert tools and similar will rely on the model. If it's out of sync there is a larger risk of issues.
> 
> It would be good to understand exactly what that concretely means... if I'm understanding you right, all we have is some additional column in the DB which isn't in the model. If the tool is relying on the model, it should simply ignore that column - just as EF does. Of course, unknown external tools can have arbitrary behavior and I can't predict how they would react, but it would be good to at least have a concrete example or two.
> 
> In summary, I'm still not sure why one would one to keep the property in the model after stopping to use it. However, if this is what you want, shadow property seem like the right answer; I'm not sure if EF needs to provide anything else here.

Interesting. I'll try out shadow properties. I have a feeling that will still be included in the query so it will not work. But I'll test and get back to you with the result.

Regarding some of the issues it will kinda work if you leave the lingering column around in the DB but remove it from code. But it's much more nicer if we can keep the DB clean of deprecated code too. In particular since the DB have no concept of marking things like obsolete like the .net code does.

**Comment by roji** [comment]
> I'll try out shadow properties. I have a feeling that will still be included in the query so it will not work.

Shadow queries *can* be included in queries, but you have to be very, very explicit about it, by using EF.Property with the correct string name for the property. This is far more explicit than just leaving a public CLR property, which anyone can reference my mistake; so I think this isn't a bad option.

> But it's much more nicer if we can keep the DB clean of deprecated code too.

Well, that is the default behavior after all - if you remove a property from your model, migrations will remove it from the database :) Your point is that you want to *keep* the column in the database for 0-downtime scenarios (e.g. for easy rollback).

> In particular since the DB have no concept of marking things like obsolete like the .net code does.

You can use comments to mark the column as obsolete, and even build various infrastructure around that (i.e. "drop all obsolete columns in my database please").

**Comment by DanielStout5** [comment]
Is there a technical reason a solution similar to below couldn't be supported?


> The proposed solution is cumbersome and a error-prone since it relies on manually editing the migration files over a long time period. Not to mention it requires 2 DB migrations to drop a single column.
> 
> 1. Remove it from the model
> 2. Add a migration
> 3. Remove the code from that migration
> 4. Deploy the update
> 5. Wait a safe amount of time (possibly weeks)
> 6. Add a new migration with no changes
> 7. Copy the removed code from the old migration in step 3
> 8. Deploy the update
> 
> I was really hoping there would be an option on EntityTypeBuilder to mark a property as excluded from queries without dropping it from the database.
> 
> I tried ignoring it e.g. `builder.Entity<MyEntity>().Ignore(d => d.ToRemove);`  but that acts just like the property was removed from MyEntity (dropping it in the next migration)
> 
> Changing it to a Shadow Property didn't drop the column, but also didn't exclude the column from queries.
> 
> ### Describe the solution you'd like
> 
> One idea is to have a method like `builder.Entity<MyEntity>().Property(x => x.ToRemove).ExcludeFromQueries();` but that could be confusing because the property will still exist on the model so there might still be code referencing it without realizing the value will not be set.
> 
> I think it would be better to have an overload of the existing Property method used for shadow properties, except with a new parameter to exclude it from queries. E.g. `builder.Entity<MyEntity>().Property<bool>("ToRemove", fetch: false)`
> 
> Then the process becomes less error prone:
> 
> 1. Remove from Model and add the DbContext line
> 2. Deploy the update
> 3. Wait safe time
> 4. Remove DbContext line
> 5. Add migration
> 6. Deploy update

**Comment by roji** [comment]
@DanielStout5 how is that different from https://github.com/dotnet/efcore/issues/27165#issuecomment-1019236727 just above?

**Comment by roji** [comment]
@DanielStout5 as written above, you can already almost achieve that, by removing the property from your CLR type, and then adding it as a shadow property:

```c#
modelBuilder.Entity<Blog>().Property<int>("SomeProperty_REMOVED").HasColumnName("SomeProperty");
```

This ensures that the property remains in EF's model representation even though there's no longer a .NET property in the actual .NET type that users interact with. The `SomeProperty` is preservedi n the database, but the only way to query the property is to very explicitly do `Where(b => EF.Property<int>("SomeProperty_REMOVED"`). Yes, a developer purposefully trying to get around the removal might do that, but nobody will do that accidentally.

It's not 100% exactly what you're asking for, but it's 99% there and it's available today.

**Comment by DanielStout5** [comment]
@roji When I did that (removed the property from the model class and made it a shadow property) the column was still fetched in all queries, so that doesn't help this at all.

**Comment by roji** [comment]
@DanielStout5 why doesn't that help? What is it important that the property not be fetched?

**Comment by MikaelEliasson** [comment]
@roji As described in the inital post it makes it very difficult to remove columns with 0-downtime deploys

> The reason is that EF generate queries with all columns named like this
> 
> ```
> SELECT [p].[Id], [p].[Name], [p].[PropToBeDeleted]
> FROM [Persons] AS [p]
> ```
> 
> So when we migrate the DB during a release the old code will try to query a column that doesn't exist (unless we do projections everywhere).

In the long run we need to be able to clean up our database too. It's just not a great approach to software engineering to keep a bunch of deprecated columns around in the database.

**Comment by roji** [comment]
I'm honestly confused here - we may be discussing different scenarios here.

Fundamentally, either you want your column to remain in the database, or you don't want it to remain in the database; if it's the second case, you can already use EF today to drop the column immediately, and everything works. So I don't think there's any need to discuss that.

So assuming you want the column to **remain** in the database, the proposed solution above (which works today) is to simply change the CLR property into a shadow property; this leaves the property in the EF model (and therefore also in the database), but at the same time removes the ability to (easily) use it from queries - developers must very explicitly use EF.Property() if they want to, which nobody will do accidentally.

For that scenario, why would it matter if it the columnj gets fetched in EF's queries, into a shadow property which nobody actually cares about? Sure, you're fetching extra data in your queries, but that's a very temporary situation (until the property really is removed), and is unlikely to matter in the grand scheme of things.

**Comment by MikaelEliasson** [comment]
No, you can't make EF drop the column if you want 0-downtime deploys. 

Let me clarify the timeline of a release in a production level system. 

```
A) Drop the property + make a migration that deletes the controller in development
B) Start a release
C) Apply the migrations to the DB (this will drop the column)
D) Swap in the new code in production 
```

In the time between C) and D) a lot of queries relating this table will be broken because any query that either tries to read or write to the now missing column is broken.

This is even worse because while the app is healthy after D) happens we run into problems if we need roll back to a previous version of the code (code be some unrelated things causing problems). 

This is why we want a solution to **Step 1: first tell EF to start treating that column as non-existing but leave it in the database** and then  **Step 2: in a later release delete it fully.**.  If we have this we still need to delete a column in two phases but at least it's doable and with two easy steps and there are no hacks involved.

Just to be clear: *Step 1* is what we are missing a simple way to do today.

**Comment by roji** [comment]
@MikaelEliasson we've been going back and forth quite a lot around this.

What you're asking for seems well-supported already by making the CLR property into a shadow property (this means application code can't use it any more). That's not a migration, since no database change occurs - the database column is preserved along with all the data, so rolling back the change is simply a matter of changing the shadow property back to a CLR property. Then, at the point where you want to really drop the column, you can remove the shadow property as well.

I'm still unclear as to what the problems with this approach are.

**Comment by AlexandreBossard** [comment]
Running across that particular issue and pretty surprised there is no proper way of doing this.

> I'm still unclear as to what the problems with this approach are.

Shadow properties does not work, because the new code (just like the old code) will continue to fetch the property you shadowed, which is weird, and technically exactly  what we don't want.

When the goal is to remove the column, you first have to make sure nobody is using it anymore. Removing it from the model is a pretty good way to enforce that. But, if you want zero downtime, during roll-out, you can't delete the doomed colomn from your DB schema as the old cold is still using it. 

That's why you need to inhibit the column suppression during migration while still instructing ef to "ignore" it during read/write. 

```
builder.Entity<MyEntity>().Property<bool>("ToRemove", fetch: false)
```

I like this proposed solution, with proper doc it would fit the use case.

**Comment by roji** [comment]
> Shadow properties does not work, because the new code (just like the old code) will continue to fetch the property you shadowed, which is weird, and technically exactly what we don't want.

Does it matter to you if EF happens to fetch the shadowed property, if that property doesn't actually exist on your .NET types? I get that from a very purist point of view you'd like it to completely disappear from the model, but I don't see what actual, concrete problem the existence of the shadow property creates.

Once again, making your property into a shadow property effectively prevents any code from using it - unless it does so very explicitly. That's extremely close to "removing it from the model", even if it's not 100% there.

**Comment by AlexandreBossard** [comment]
It's not a purist point, it's a "it does not help in a shape or form"
point.

If you deploy your new code that uses shadow property, your service still
fetches the property under the hood. Then you cannot delete the column in a
follow up migration, and you are back to square one.

Or am I missing something ?

Le mar. 20 mai 2025, 18:13, Shay Rojansky ***@***.***> a
écrit :

**Comment by roji** [comment]
> Or am I missing something ?

I think you are.

Phase 1: Introduce a new property, change the old into a shadow property. Your application code is changed to only use the new property. During this phase, the old property is still in the EF model (in shadow state), and as a result continues to exist as a column in the database (no migration is generated to drop it). Yes, your EF queries fetch the old column, but application code can't actually use that without very explicitly looking for the shadow property values (e.g. in the change tracker). Again, barring explicit access of shadow state, as far as your application is concerned the property is gone. This state is not 100% perfect, but it IMHO addresses the main need: the column isn't dropped, but it's also not in active use any more; it's very easy to switch back by making the old property non-shadow again etc.

Phase 2: remove the old shadow property, generate a migration which drops the old column from the database.

**Comment by bonbor** [comment]
> Phase 2: remove the old shadow property, generate a migration which drops the old column from the database.

How are we going to deploy this without down time? The running version of the application is still generating and executing SQL which depend on the shadow property. When the migration is executed, the column is removed and the SQL statements will fail.

**Comment by roji** [comment]
Phase 2 is done only when you've fully transitioned to the new property/column, are happy with the new state of affairs, and want to delete the old property/column. No actual application code is doing anything with it at that point.

I recommend reading all the comments above - including from before your first post, that might shed more light on what I'm trying to explain.

**Comment by AlexandreBossard** [comment]
So I've fooled around with some basic examples. And I'm pretty sure that the only thing a shadow property provide is:
1. Remove the use of the property dropped through your code
2. Not generating a drop column migration

But as long as you have any version of you old service running with either the model property or the shadow property, EF will select it unless you do manual projection. 

A simple
```csharp
app.MapGet("/Blog", (BloggingContext ctx) =>
{
  return ctx.Blogs;
})
```
will generate 
```fail: Microsoft.EntityFrameworkCore.Database.Command[20102]
      Failed executing DbCommand (3ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT "b"."BlogId", "b"."Description", "b"."Url"
      FROM "Blogs" AS "b"
fail: Microsoft.EntityFrameworkCore.Query[10100]
```
as long you have either two commented lines, uncommented. 

```csharp
public class Blog
{
  public Guid BlogId { get; set; }
  public required string Url { get; set; }
  // public string? Description { get; set; } = null;

  public List<Post> Posts { get; } = [];
}
public class BloggingContext : DbContext
{
  public BloggingContext(DbContextOptions<BloggingContext> options)
    : base(options)
  { }
  protected override void OnModelCreating(ModelBuilder modelBuilder)
  {
    // modelBuilder.Entity<Blog>().Property<string>("Description");
  }
  public DbSet<Blog> Blogs { get; set; }
  public DbSet<Post> Posts { get; set; }
}
```
So we need a way to tell EF to stop selecting the column / ignore it while not dropping it during a migration.

Step 0: Original state
Step 1: Remove the column from the model while keeping it in the DB schema, no migration generated
State 2: Progressively roll out new version without down time, letting old version sunset out while using old property/column without error.
Step 3: Generate a migration that drops the column
Step 4: Deploy migration, without downtime
Step 5: Go for lunch.

**Comment by AlexandreBossard** [comment]
Yes we are and will go in circles unless you realize the crux of the
problem is that EF will fetch the property. You don't seem to get that when
you do zero down time deployment of a multi instances service, you have 2
versions (old/new) cohabiting on the same database. This is why you can't drop
the column.

Advocating for keeping columns, disregarding fetching through shadow
property is silly at best. Also advocating for an undocumented trick, that
is breaking philosophy of EF is also concerning.

Why not acknowledge the proposed solution above? Or at least document the
ones you propose in the official documentation ?


Le mer. 21 mai 2025, 22:45, Shay Rojansky ***@***.***> a
écrit :

**Comment by roji** [comment]
> You don't seem to get that when you do zero down time deployment of a multi instances service, you have 2 versions (old/new) cohabiting on the same database.

I do get that - that's exactly the setup I've been discussing here: keep both columns in the database, while not using the old column in the (new) EF application.

> This is why can't drop the column.

And I never said the column is dropped - until everything has been migrated to the new column and the old one can be safely removed. That's what we've been discussing from the start.

> Advocating for keeping columns, disregarding fetching through shadow property is sealy at best.

I'm sorry, but I simply don't understand. I'm not advocating for keeping columns; keeping the old column - temporarily, during the transition phase - is the whole point of the discussion. If you want to immediately drop the column, just do the regular EF flow, removing the property from the model and then immediately having the corresponding column removed as well.

> Also advocating for an undocumented trick, that is breaking philosophy of EF is also concerning.

There is no undocumented trick here, and I have no idea what "philosophy of EF" this breaks. Shadow properties are an official, 1st-class concept in EF, and I'm proposing using it as a way of hiding the column from the application during the transitional phase, when the new column should be used instead.

In any case, I think this conversation has run its course. There are already several valid techniques proposed above for dealing with this scenario - you may choose whatever fits your needs. I agree that we'd ideally have a page guiding people on how to do this, but it would unfortunately rank quite low in the current priority of things (there are more important things needing documentation). But please feel free to submit an issue to https://github.com/dotnet/entityFramework.Docs to make sure it's tracked.

**Comment by stevendarby** [comment]
Sorry if it's unhelpful to add another reply at this point, but I really feel like you're reading past a lot of the replies here @roji and I wanted one last go at trying to make it clearer. 

A single deployment typically works by first applying the database migrations and then upgrading the service. It's the gap between the two - which might only be minutes or even seconds - which you need to consider carefully. 

It means that you cannot remove a column in a single deployment because once the migration is run, there is a short period where the previous version of the service is still running and including that column in queries against a new DB schema that no longer has it. 

Therefore the approach is usually:

1. First deploy a change to the service so that the column is no longer used _at all_. 
2. Do a second deployment with a migration to drop the column from the DB. 

You cannot achieve step 1 by making it a shadow property because it's still going to be included in full entity queries - this is outside of the developers control.

**Comment by roji** [comment]
@stevendarby thanks for stepping in, but I'm still missing something - help me understand. To my understanding, the point of the 1st step is to **keep** the old column in the database, but have (new deployments) no longer use it (because they've switched to using another column). Or as @ajcvickers summarized at the top of the issue ([comment](https://github.com/dotnet/efcore/issues/27165#issuecomment-1012940910)):

> [...] can I clarify that the property is no longer used by newer versions of the application, but that you want the database column to be retained so that older versions of the application do not break?

Assuming that's what we're talking about... Making the old property into a shadow property means that EF indeed continues to query for it in queries. However, given that there's no longer a CLR property on your CLR type, the application no longer has (easy) access to it. All places where the old property was used no longer compile (the .NET property is no longer there), and the developer now needs to change them to reference the new property, which is what we want.

As I wrote several times above, this isn't a 100% "pure" solution: a dev can still access the shadow property via the change tracker (or the column via EF.Property), but those are very explicit actions that presumably nobody will do unintentionally. Also, as you point out, EF will continue to include the old column in its queries - but I'm still struggling to understand why that is a problem; why should a developer care that an unused column is being fetched from the database, as long as that data is pretty much inaccessible to application code? Yes, there's a slight performance hit in the extra column fetch, but in the vast majority of cases that shouldn't matter.

Help me understand what I'm missing here.

**Comment by havarnov** [comment]
@roji I think what your missing is that a lot of us run migration before we do a code update.

t1: migrate database
t2: update code

deltaT = t2 - t1, might be minutes or seconds, but in this period you have code that in effect does `SELECT X FROM Entity` regardless if it's a shadow copy or from the CLR type itself. But since X is dropped at t1, the code (old code) will fail.

What we need is to do something like:

t1: make EF _not_ query for property X
t2: run migration (DROP COLUMN X ...)
t3: update code

**Comment by roji** [comment]
@havarnov what you're describing is very different from what was originally asked in this issue - that's fine in itself, but we seem to be conflating multiple scenarios here, leading to lots of confusion.

Can you please provide a full, step-by-step description of what you're trying to do, to be sure we're talking and thinking about the same thing?

In your above example, the phase between t1 and t2 still has the column in the database, but you don't want your application to use it: if that's the case, that's where I propose using shadow properties. Note that your application not using X is different from EF not querying X, but during this phase the column is still in the database, so that shouldn't matter.

Then, once you reach t2, you want to really get rid of the column; at that point, you remove the shadow property from the EF model and generate a migration which drops the corresponding column. That's just standard EF migrations.

**Comment by stevendarby** [comment]
@roji

Let me start my being clear that we are NOT concerned about the column being fetched because of the extra data... this is trivial. We're concerned because _it will not exist in the database_!

Next thing to be clear on is that migration and service upgrade are not simultaneous. Typically the database is upgraded and then the service is upgraded. This means there is a period in between the two where the database schema is version _n_ and the service is version _n-1_.

Let's walk through an example. Version 1 is the starting point where the column exists in the model and in the database and is actively used in developer written queries. 

We then make the property a shadow property and ensure it's never explicitly used in code. This is your suggested first step. This is version 2. We happily deploy this version with no hiccups. Although there were no migrations, we still ran through the migration process (which found nothing to do), so for ease let's just say both database and service are now on version 2.

In version 2, the service is still querying for the shadow property's column whenever the entity is queried.

Next, we fully remove the shadow property too, producing a migration to remove the column. This is version 3. We deploy version 3. The first thing we do is run the migrations. This completes successfully. The database is now on version 3 (no column) but the service is on version 2 (shadow property still being queried). Requests are made to the service during this period - the service errors because it tries to load the entity using a column that doesn't exist!

The deployment continues on to upgrade the service. Now both database and service are on version 3, so we've recovered, but we had a short period where things failed and customers are unhappy.

Ultimately, you need to be able to produce a version where the column queried _at all_. Only after you've deployed that you can do another deployment to drop the column as both version n _and_ n-1 aren't querying it.

**Comment by stevendarby** [comment]
@roji

> @havarnov what you're describing is very different from what was originally asked in this issue - that's fine in itself, but we seem to be conflating multiple scenarios here, leading to lots of confusion.

Sorry but it's exactly what the original post was about. Zero downtime deployments. Take this snippet:

> So when we migrate the DB during a release the old code will try to query a column that doesn't exist (unless we do projections everywhere).

The old code they refer to is that n-1 version that is still live briefly after migrations are run.

**Comment by stevendarby** [comment]
Final comment, personally I think this is a fine workaround: https://github.com/dotnet/efcore/issues/27165#issuecomment-1019236727

After that has been deployed, you can then produce a new blank migration, add in a drop column statement and release that. 

Others have since asked for a more first class solution. You can disagree with that but not with the fact that it solves a real problem.

**Comment by MikaelEliasson** [comment]
I think that hack will work even if I wish we had a better way to do it. I suspect it could cause issues for some tools / scenarios if the EF model and the DB is out of sync but that should be rare. And if people run into that case it should be possible to revisit the decision then.

I would suggest for people to do an Extension method that just wraps `Ignore`

```
x.Entity<MyEntity>().IgnoreForDeletion(x => x.PropToDelete, marked: dateTimeThisWasAdded);
```

This way you at least have a chance to find all those props you started to delete but forgot to finish... It will ofc leave the prop in the C# code a bit longer. But that is the smaller issue as you can always mark it as [Obsolete]

**Comment by MikaelEliasson** [comment]
> Our CI checks for model desync through ‘dotnet ef has-pending-changes‘. I would have to make a workaround for the workaround.

Unless I'm mistaking that only checks if the application model and the snapshot is out of sync. It does not compare against the database. So you should be fine.

**Comment by blastrock** [comment]
Hi,

I think we all agree on the goal, but we are clearly not on the same page.

@roji, either we didn't understand what you propose, or you didn't understand our issue with it. So, please correct me if I'm wrong. You acknowledged that:

- Having shadow properties will still make EF fetch and insert the columns into the tables
- You propose that as a temporary measure, the goal being to eventually drop the column with zero downtime

Do we agree on this?

I'll try and describe what you propose in details to understand where we disagree.

Let's assume we have a table

```sql
CREATE TABLE t (id int, to_drop int);
```

As @havarnov said, common deployment pipelines have 2 steps: run the database migration, and then update the running services. So to drop a column according to what you propose, it would go like this:

1. Mark `to_drop` as a shadow property
    a. It'll still be fetched and inserted
    b. No migration will be generated
2. Deploy this new server
3. Drop the column from the model, generate a migration
4. Deploy this new server
    a. The CI will first run the migrator and run `DROP COLUMN to_drop`
    b. The old servers that are still running with the shadow property will continue to run `SELECT id, to_drop FROM t`. This query will fail because `to_drop` does not exist anymore, and this is our issue.
    c. The new server will be deployed, it won't select or insert the `to_drop` column anymore, everything is in sync

To clarify: our goal is to be able to run this scenario without any server error. It's fine if the servers are fetching extra data temporarily, the issue is that the server must not throw 500s during the deployment phase. So the flow I just described does not work for us.

How can we avoid server errors during this migration?

**Comment by DanielStout5** [comment]
> [@DanielStout5](https://github.com/DanielStout5) why doesn't that help? Why is it important that the property not be fetched?

I think this perhaps has been answered by other folks above, but just to clarify: It's important that the property not be fetched because the code needs to work regardless of whether the column exists in the database. If EF core is still attempting to fetch it after the column has been removed, then the application will throw an exception.

We need to be able to "hide" the existence of the column from EF Core **before** we actually delete it from the database - after all servers are running that code, we can actually delete the column and then fully remove any reference to it in the code.

**Comment by KaliCZ** [comment]
Since it hasn't been acknowledged as a real problem, I'm hoping I could summarize why exactly it is a problem to majority of applications and not some edge-case that can be disregarded. 

# Example case
Imagine you have a SQL database and a web app connecting to that database. Presumably you can have multiple slots such as a staging slot. Presumably you can even have multiple app services connecting to a single database.

So what happens if you have an unused property in EF model that you want to remove and also an unused column in the database that you want to remove, because we care about our database just as much as our code? 

## The way to solve this
Well you certainly can't just remove it and create a migration that does `drop column` and execute this migration. That would break all the other servers connecting to our database - not all servers get deployed at the same time. You need some grace period. 

What is this grace period? Grace period is a period during which you stop using the column which will then make it safe to be deleted in the future. So your first deploy must be simply unmapping this property without affecting the database structure. But that unmapping must make sure that the property is safe to be removed. Therefore it must not be fetched with any type of query.

Then in your second deploy, you can execute the breaking migration `drop column`. Only after you have made sure that removing this column won't break any application.

So the process for deployment is:
1. Have a property in the model + column in DB
2. Remove a property from the model, keep column in DB, do not use it
3. Remove the column from DB, because it's no longer used.

You need one deploy to move from 1 to 2 and another deploy to move from 2 to 3. And shadow properties don't move you from 1 to 2. That's the problem @roji 


To address your questions:
-----
> Fundamentally, either you want your column to remain in the database, or you don't want it to remain in the database; if it's the second case, you can already use EF today to drop the column immediately, and everything works. So I don't think there's any need to discuss that.

You can't simply remove the column - not when you have more than 1 server. The moment that property is removed, all your servers stop being able to query your database until they receive the updated code.

> For that scenario, why would it matter if it the columnj gets fetched in EF's queries, into a shadow property which nobody actually cares about?

We don't care that the property is fetched. We care that entity framework suddenly throws exception on all queries that get executed when the column suddenly disappears. I don't want to receive `Invalid column name 'UnusedColumn'.` because of something that's not supposed to be used.

Summary
-----
Please, stop thinking about 1 database and 1 server where you can simply execute breaking migrations. That's not the modern world of cloud services. Any basic scaling means you have more than 1 server, so you cannot execute a breaking migration without multiple deploys. Even the current recommendation is to build `efbundle` and execute it in the pipeline. That means your `drop column` gets executed before your code reaches production. Therefore you have broken your production server, There needs to be a possibility to truly unmap a property for it to be removed in the second deploy.

I hope I've made it clear what the problem is. If not, I'm available for a follow-up or a phone call.

Existing solution
======
For people that arrive to this thread late, I feel the need to mention that there is an existing solution mentioned above. We just don't like it because it's cumbersome. Here's how to do it currently:
1. Remove a property from your model and generate a migration
2. Comment the entire up and down methods of this migration. This migration doesn't actually do anything, it only unmaps the property from your queries
3. Deploy
4. Create a second migration, put the commented code from the first migration into the second migration (and uncomment it 😄 ) 
5. Deploy
6. 🎉 You're done

**Comment by ocoster-sermo** [comment]
The ideal behaviour for me would be:

Marking a property as `[Obsolete]` would cause it to no longer be fetched from the database. It would not remove it from the model (so a new migration would be empty, as it would currently).

This would make the code safe to deploy (and in codebases where warnings are treated as errors, the property would be certain to not be used by anything).

At this point, removing the property is possible and and a new migration would remove it from the model (again, as it would currently).

---

I appreciate this is likely a breaking change, so may be something to implement with an opt-in flag.

The alternative, is to have a different mechanism to opt-out properties from being fetched from the DB (i.e., most of the discussion here).
